testplan/common/entity/base.py

```diff
@@ -188,6 +188,14 @@ def all_status(self, target):
             for resource in self._resources
         )
 
+    def _fetch_log_messages(self, resource, message):
+        if hasattr(resource, 'fetch_error_log'):
```

Comment:
if you expect content of the environment to be Resources and that class has implementation of fetch_error_log you do not need the check. If you enable non Resources, then I suggest to go with Protocol as a much modern alternative.

Reply:
I am maintaining the class hierarchy implementation.